upload/src/addons/SV/WarningImprovements/XF/ControllerPlugin/Warn.php

```diff
@@ -2,11 +2,38 @@
 
 namespace SV\WarningImprovements\XF\ControllerPlugin;
 
+use XF\Mvc\Entity\Entity;
+
 /**
  * Extends \XF\ControllerPlugin\Warn
  */
 class Warn extends XFCP_Warn
 {
+    public function actionWarn($contentType, Entity $content, $warnUrl, array $breadcrumbs = [])
+    {
+        $response = parent::actionWarn($contentType, $content, $warnUrl, $breadcrumbs);
+
+        if ($response instanceof \XF\Mvc\Reply\View)
+        {
+            $categoryRepo = $this->getWarningCategoryRepo();
+            $categoryTree = $categoryRepo->createCategoryTree(null, 0, true);
+
+            /** @var \XF\Repository\Warning $warningRepo */
+            $warningRepo = $this->repository('XF:Warning');
+            $warnings = $warningRepo->findWarningDefinitionsForList()
+                ->order('sv_display_order', 'asc')
+                ->fetch()
+                ->groupBy('sv_warning_category_id');
```

Comment:
I think these bits can be pushed into `findWarningDefinitionsForList()`, as every case they are used by this add-on these arguments are added to `findWarningDefinitionsForList()` and XF2 just passes the warning list around.

Means the warning list is only fetched once.

Reply:
Done

Reply:
On second added a new method named `findWarningDefinitionsForListGroupedByCategory` because XF considers `findWarningDefinitionsForList` to return a Finder